bin/git-cp

```diff
@@ -0,0 +1,64 @@
+#!/usr/bin/env bash
+set -euo pipefail
+
+PROGRAM=$0
+CURRENT_FILENAME=""
+DESTINATION_FILENAME=""
+
+function usage()
+{
+    echo 1>&2 "USAGE: ${PROGRAM} CURRENT_FILENAME DESTINATION_FILENAME"
+}
+
+while [[ $# -gt 0 ]]
+do
+  key="$1"
+  
+  if [[ "$CURRENT_FILENAME" == "" ]]; then
+      CURRENT_FILENAME=$key
+  elif [[ "$DESTINATION_FILENAME" == "" ]]; then
+      DESTINATION_FILENAME=$key
+  else
+      usage
+      exit 30  # Error during arguments parsing
+  fi
+  shift  # past argument or value
+done
+
+if [[ "$DESTINATION_FILENAME" == "" ]]; then
+    usage
+    exit 20  # Missing arguments CURRENT_FILENAME
+elif [[ "$CURRENT_FILENAME" == "" ]]; then
+    usage
+    exit 10  # Missing arguments CURRENT_FILENAME
+else
+    if [ ! -e "$DESTINATION_FILENAME" ]; then
+        echo 1>&2 "$DESTINATION_FILENAME already exists."
+        echo 1>&2 "Make sure to remove the destination first."
+        exit 40
+    fi
+    
+    echo "Coping $CURRENT_FILENAME into $DESTINATION_FILENAME"
+    INTERMEDIATE_FILENAME="${CURRENT_FILENAME//\//__}-move-to-${DESTINATION_FILENAME//\//__}"
+
+    # We keep the existing file on the side in a commit
+    git mv "${CURRENT_FILENAME}" "${INTERMEDIATE_FILENAME}"
+    git commit -nm "Keep $CURRENT_FILENAME"
+    SAVED=$(git rev-parse HEAD)
+
+    # We come back to the previous state
+    git reset --hard HEAD^  # Revert that change
+
+    # We move the file to its new destination
+    git mv "${CURRENT_FILENAME}" "${DESTINATION_FILENAME}"
+    git add "${DESTINATION_FILENAME}"
+    git commit -nm "Copy $CURRENT_FILENAME into $DESTINATION_FILENAME"
+
+    # We get back our copy
+    git merge "${SAVED}"
+    git commit -a -m "Duplicate ${CURRENT_FILENAME} history."  # Keeping both files
+
+    # We get back our original name
+    git mv "${INTERMEDIATE_FILENAME}" "${CURRENT_FILENAME}"
+    git commit -nm "Set back $CURRENT_FILENAME file."
```

Comment:
@Natim
Maybe we can just leave one commit as the history to keep? Create four new commits per operation is too scary to me.

Reply:
If you figure out a way to reduce the number of commit I would be interested indeed.

Reply:
It might be possible using git plumbering

Reply:
All we have to do is to create a new object with the same parent object.

Reply:
> **osse on #git-devel IRC channel**:
> file objects as in blobs? they don't have parents

Reply:
Refs https://stackoverflow.com/a/1043566/186202